src/fmt.rs

```diff
@@ -187,58 +179,109 @@ const LOWER: [u8; 16] = [
-/// group.
-const HYPHEN_POSITIONS: [usize; 4] = [8, 13, 18, 23];
-
-/// Encodes the `uuid` possibly with hyphens, and possibly in upper
-/// case, to full_buffer[start..] and returns the str sliced from
-/// full_buffer[..start + encoded_length].
-///
-/// The `start` parameter allows writing a prefix (such as
-/// "urn:uuid:") to the buffer that's included in the final encoded
-/// UUID.
-#[allow(clippy::needless_range_loop)]
-fn encode<'a>(
-    full_buffer: &'a mut [u8],
-    start: usize,
-    trailing: usize,
-    uuid: &Uuid,
-    hyphens: bool,
-    upper: bool,
-) -> &'a mut str {
-    let len = if hyphens { 36 } else { 32 };
-
-    {
-        let buffer = &mut full_buffer[start..start + len];
-        let bytes = uuid.as_bytes();
-
-        let hex = if upper { &UPPER } else { &LOWER };
-
-        for group in 0..5 {
-            // If we're writing hyphens, we need to shift the output
-            // location along by how many of them have been written
-            // before this point. That's exactly the (0-indexed) group
-            // number.
-            let hyphens_before = if hyphens { group } else { 0 };
-            for idx in BYTE_POSITIONS[group]..BYTE_POSITIONS[group + 1] {
-                let b = bytes[idx];
-                let out_idx = hyphens_before + 2 * idx;
-
-                buffer[out_idx] = hex[(b >> 4) as usize];
-                buffer[out_idx + 1] = hex[(b & 0b1111) as usize];
-            }
 
-            if group != 4 && hyphens {
-                buffer[HYPHEN_POSITIONS[group]] = b'-';
-            }
+#[inline]
+const fn format_simple(src: &[u8; 16], upper: bool) -> [u8; 32] {
+    let lut = if upper { &UPPER } else { &LOWER };
+    let mut dst = [0; 32];
+    let mut i = 0;
+    while i < 16 {
+        let x = src[i];
+        dst[i * 2] = lut[(x >> 4) as usize];
+        dst[i * 2 + 1] = lut[(x & 0x0f) as usize];
+        i += 1;
+    }
+    dst
+}
+
+#[inline]
+const fn format_hyphenated(src: &[u8; 16], upper: bool) -> [u8; 36] {
+    let lut = if upper { &UPPER } else { &LOWER };
+    let groups = [(0, 8), (9, 13), (14, 18), (19, 23), (24, 36)];
+    let mut dst = [0; 36];
+
+    let mut group_idx = 0;
+    let mut i = 0;
+    while group_idx < 5 {
+        let (start, end) = groups[group_idx];
+        let mut j = start;
+        while j < end {
+            let x = src[i];
+            i += 1;
+
+            dst[j] = lut[(x >> 4) as usize];
+            dst[j + 1] = lut[(x & 0x0f) as usize];
+            j += 2;
+        }
+        if group_idx < 4 {
+            dst[end] = b'-';
         }
+        group_idx += 1;
     }
+    dst
+}
 
-    str::from_utf8_mut(&mut full_buffer[..start + len + trailing])
-        .expect("found non-ASCII output characters while encoding a UUID")
+#[inline]
+fn encode_simple<'b>(
+    src: &[u8; 16],
+    buffer: &'b mut [u8],
+    upper: bool,
+) -> &'b mut str {
+    const LEN: usize = 32;
+    let buf = &mut buffer[..LEN];
+    unsafe {
```

Comment:
Maybe we could move the SAFETY comment to outside of the `unsafe` block here to also cover the `ptr::write` call?

The `as_mut_ptr` call isn't unsafe so we could also pull that one out.